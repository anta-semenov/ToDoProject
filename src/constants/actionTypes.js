/*
 * Project
 */
export const ADD_PROJECT = 'ADD_PROJECT'
export const REMOVE_PROJECT = 'REMOVE_PROJECT'
export const COMPLETE_PROJECT = 'COMPLETE_PROJECT'
export const EDIT_PROJECT = 'EDIT_PROJECT'
export const REPLACE_PROJECT = 'REPLACE_PROJECT'

/*
 * Task
 */
export const ADD_TASK = 'ADD_TASK'
export const REMOVE_TASK = 'REMOVE_TASK'
export const COMPLETE_TASK = 'COMPLETE_TASK'
export const EDIT_TASK = 'EDIT_TASK'
export const REPLACE_TASK = 'REPLACE_TASK'
export const ADD_TASK_TO_PROJECT = 'ADD_TASK_TO_PROJECT'
export const ADD_TASK_CONTEXT = 'ADD_TASK_CONTEXT'
export const REMOVE_TASK_CONTEXT = 'REMOVE_TASK_CONTEXT'
export const SET_TASK_TODAY = 'SET_TASK_TODAY'
export const START_TASK_TRACKING = 'START_TASK_TRACKING'
export const STOP_TASK_TRACKING = 'STOP_TASK_TRACKING'

/*
 * Context
 */
export const ADD_CONTEXT = 'ADD_CONTEXT'
export const REMOVE_CONTEXT = 'REMOVE_CONTEXT'
export const EDIT_CONTEXT = 'EDIT_CONTEXT'
export const SWITCH_TASK_CONTEXT = 'SWITCH_TASK_CONTEXT'
export const REPLACE_CONTEXT = 'REPLACE_CONTEXT'

/*
* UI State
*/
export const SET_SELECTED_SECTION = 'SET_SELECTED_SECTION'
export const SET_SIDEBAR_SIZE = 'SET_SIDEBAR_SIZE'
export const SET_ACTIVE_ITEM = 'SET_ACTIVE_ITEM'
export const SET_EDITING_SECTION = 'SET_EDITING_SECTION'
export const TOGGLE_TASK_COMPLETED_LATENCY = 'TOGGLE_TASK_COMPLETED_LATENCY'
export const CLEAR_COMPLETED_LATENT_TASKS = 'CLEAR_COMPLETED_LATENT_TASKS'
export const TOGGLE_TASK_LATENCY = 'TOGGLE_TASK_LATENCY'
export const CLEAR_LATENT_TASKS = 'CLEAR_LATENT_TASKS'
export const SET_SYNCING = 'SET_SYNCING'
export const SET_OFFLINE = 'SET_OFFLINE'
export const SET_AUTH_STATUS = 'SET_AUTH_STATUS'
export const SET_AUTH_ERROR_MESSAGE = 'SET_AUTH_ERROR_MESSAGE'
export const SET_PROPERTY = 'SET_PROPERTY'

/*
* Common
*/
export const SET_STATE = 'SET_STATE'
export const LOG_OUT = 'LOG_OUT'
export const REQUEST_AUTH = 'REQUEST_AUTH'
export const RECIEVE_AUTH = 'RECIEVE_AUTH'
export const ERROR_AUTH = 'ERROR_AUTH'

export const REQUEST_DATA = 'REQUEST_DATA'
export const RECIEVE_DATA = 'RECIEVE_DATA'
export const ERROR_DATA = 'ERROR_DATA'

export const UNDO = 'UNDO'
export const REDO = 'REDO'
export const CLEAR_UNDO_REDO = 'CLEAR_UNDO_REDO'
