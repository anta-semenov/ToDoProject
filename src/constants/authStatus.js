export const AUTH_IN_PROGRESS = 'AUTH_IN_PROGRESS'
export const AUTH_SUCESS = 'AUTH_SUCESS'
export const AUTH_ERROR = 'AUTH_ERROR'
export const AUTH_NONE = 'AUTH_NONE'
