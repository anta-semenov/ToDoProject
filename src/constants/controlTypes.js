export const TASK_TITLE_FIELD = 'TASK_TITLE_FIELD'
export const TASK_DESCRIPTION_FIELD = 'TASK_DESCRIPTION_FIELD'
