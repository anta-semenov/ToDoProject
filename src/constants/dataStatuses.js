export const DATA_NONE = 'DATA_NONE'
export const DATA_REQUESTED = 'DATA_REQUESTED'
export const DATA_ERROR = 'DATA_ERROR'
export const DATA_RECIEVED = 'DATA_RECIEVED'
