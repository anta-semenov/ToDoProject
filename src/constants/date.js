import { fromJS } from 'immutable'
export const WEEKDAY_NAMES = fromJS([
  'Sunday',
  'Monday',
  'Tuesday',
  'Wednesday',
  'Thursday',
  'Friday',
  'Saturday'
])
export const WEEKDAY_SHORT_NAMES = fromJS([
  'Su',
  'Mo',
  'Tu',
  'We',
  'Th',
  'Fr',
  'Sa'
])

export const MONTH_NAMES = fromJS([
  'January',
  'February',
  'March',
  'April',
  'May',
  'June',
  'July',
  'August',
  'September',
  'October',
  'November',
  'December'
])
export const MONTH_SHORT_NAMES = fromJS([
  'Jan',
  'Feb',
  'Mar',
  'Apr',
  'May',
  'Jun',
  'Jul',
  'Aug',
  'Sep',
  'Oct',
  'Nov',
  'Dec'
])
