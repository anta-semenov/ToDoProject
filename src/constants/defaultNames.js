export const NEW_CONTEXT_TITLE = 'New Context'

export const NEW_TASK_TITLE = 'New Task'
