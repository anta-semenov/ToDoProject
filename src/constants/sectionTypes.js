export const PROJECTS = 'PROJECTS'
export const CONTEXTS = 'CONTEXTS'
export const PROJECT = 'PROJECT'
export const CONTEXT = 'CONTEXT'
export const INBOX = 'INBOX'
export const TODAY = 'TODAY'
export const NEXT = 'NEXT'
export const SOMEDAY = 'SOMEDAY'
