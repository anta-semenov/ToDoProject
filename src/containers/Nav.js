import Navigation from '../components/navigation/Navigation'
import { connect } from 'react-redux'
import { setSelectedSection, setEditingSection, clearLatentTasks } from '../actions/uiStateActions'
import { addProject, editProject, changeProjectPosition } from '../actions/projectActions'
import { addContext, editContext, changeContextPosition } from '../actions/contextActions'
import { BASIC, PROJECTS, CONTEXTS} from '../constants/navGroupTypes'
import { fromJS } from 'immutable'
import * as sectionTypes from '../constants/sectionTypes'
import * as sectionNames from '../constants/sectionNames'
import uniqueKey from '../utils/uniqueKeyGenerator'
import { ADD_NEW_CONTEXT_TITLE, ADD_NEW_PROJECT_TITLE } from '../constants/defaults'
import { getOrderedProjectsList, getOrderedContextsList } from '../reducer'

export const mapStateToProps = (state) => {
  const selectedSectionType = state.getIn(['uiState', 'selectedSection', 'type'])
  const selectedSectionID = state.getIn(['uiState', 'selectedSection', 'id'])
  const editingSectionType = state.getIn(['uiState', 'editingSection', 'type'])
  const editingSectionID = state.getIn(['uiState', 'editingSection', 'id'])

  const groups = [
    {
      type: BASIC,
      items: fromJS([
        {
          type: sectionTypes.INBOX,
          title: sectionNames.INBOX,
          active: selectedSectionType === sectionTypes.INBOX ? true : false,
          count: state.get('task').filter(task => !task.get('completed') && !task.get('today') && !task.get('someday') && !task.has('project') && !task.has('contexts')).size
        },
        {
          type: sectionTypes.TODAY,
          title: sectionNames.TODAY,
          active: selectedSectionType === sectionTypes.TODAY ? true : false,
          count: state.get('task').filter(task => !task.get('completed') && task.get('today')).size
        },
        {
          type: sectionTypes.NEXT,
          title: sectionNames.NEXT,
          active: selectedSectionType === sectionTypes.NEXT ? true : false
        },
        {
          type: sectionTypes.SOMEDAY,
          title: sectionNames.SOMEDAY,
          active: selectedSectionType === sectionTypes.SOMEDAY ? true : false
        }
      ])
    },
    {
      type: CONTEXTS,
      title: sectionNames.CONTEXTS,
      addNewTitle: ADD_NEW_CONTEXT_TITLE,
      items: getOrderedContextsList(state).map((item, index, list) => {
        const id = item.get('id')
        return fromJS({
          id: id,
          type: sectionTypes.CONTEXT,
          title: item.get('title'),
          active: selectedSectionType === sectionTypes.CONTEXT && selectedSectionID === id ? true : false,
          editing: editingSectionType === sectionTypes.CONTEXT && editingSectionID === id ? true : false,
          count: state.get('task').filter(task => !task.get('completed') && task.get('context', fromJS([])).has(id)).size,
          nextId: list.getIn([index+1,'id'])
        })
      })
    },
    {
      type: PROJECTS,
      title: sectionNames.PROJECTS,
      addNewTitle: ADD_NEW_PROJECT_TITLE,
      items: getOrderedProjectsList(state).map((item, index, list) => {
        const id = item.get('id')
        return fromJS({
          id: id,
          type: sectionTypes.PROJECT,
          title: item.get('title'),
          active: selectedSectionType === sectionTypes.PROJECT && selectedSectionID === id ? true : false,
          editing: editingSectionType === sectionTypes.PROJECT && editingSectionID === id ? true : false,
          nextId: list.getIn([index+1,'id'])
        })
      })
    }
  ]
  return {
    groups: groups
  }
}

export const mapDispatchToProps = (dispatch) => {
  return {
    onItemClick: (type, id) => {
      dispatch(clearLatentTasks())
      dispatch(setSelectedSection({type: type, id: id}))
    },
    addNew: (type) => {
      const newKey = uniqueKey()
      switch (type) {
        case sectionTypes.PROJECTS:
          dispatch(addProject({id: newKey}))
          dispatch(setEditingSection({
            type: sectionTypes.PROJECT,
            id: newKey
          }))
          break
        case sectionTypes.CONTEXTS:
          dispatch(addContext({id: newKey}))
          dispatch(setEditingSection({
            type: sectionTypes.CONTEXT,
            id: newKey
          }))
          break
      }
    },
    onStopEditing: (item) => {
      dispatch(setEditingSection())
      if (item.newTitle) {
        const properties = {title: item.newTitle}
        switch (item.type) {
          case sectionTypes.PROJECT:
            dispatch(editProject(item.id, properties))
            break
          case sectionTypes.CONTEXT:
            dispatch(editContext(item.id, properties))
            break
        }
      }
    },
    changePosition: (type, id, nextId) => {
      switch (type) {
        case sectionTypes.PROJECT:
          dispatch(changeProjectPosition(id, nextId))
          break
        case sectionTypes.CONTEXT:
          dispatch(changeContextPosition(id, nextId))
          break
      }
    }
  }
}

const Nav = connect(
  mapStateToProps,
  mapDispatchToProps
)(Navigation)

export default Nav
