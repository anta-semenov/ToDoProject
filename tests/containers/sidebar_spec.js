import { mapStateToProps } from '../../src/containers/Nav'
import { fromJS, Set } from 'immutable'
import * as sectionTypes from '../../src/constants/sectionTypes'

describe('Sidebar container', () => {
  const testState = fromJS({
    task: {
      b41sogy3s0oc: {
        id: 'b41sogy3s0oc',
        title: 'test task 1',
        completed: false,
        completedDeleted: false,
        today: false
      },
      b41sogy3s0ok: {
        id: 'b41sogy3s0ok',
        title: 'test task 2',
        completed: false,
        completedDeleted: false,
        today: true,
        project: 'ab6sof83s1o1'
      },
      b41sogy3s0om: {
        id: 'b41sogy3s0om',
        title: 'test task 3',
        completed: true,
        completedDeleted: true,
        today: false,
        project: 'ab6sof83s1o1'
      },
      b41sogy3s0oq: {
        id: 'b41sogy3s0oq',
        title: 'test task 4',
        completed: false,
        completedDeleted: false,
        today: true,
        project: 'ab6sof83s1o2',
        context: Set(['a23sogy3s0oq'])
      }
    },
    context: {
      a21sogy3s0oq: {
        id: 'a21sogy3s0oq',
        title: 'test context 1',
        deleted: false
      },
      a23sogy3s0oq: {
        id: 'a23sogy3s0oq',
        title: 'test context 2',
        delted: false
      }
    },
    project: {
      ab6sof83s1oq: {
        id: 'ab6sof83s1oq',
        title: 'test project 1',
        completed: true,
        completedDeleted: true
      },
      ab6sof83s1o1: {
        id: 'ab6sof83s1o1',
        title: 'test project 2',
        completed: false,
        completedDeleted: false
      },
      ab6sof83s1o2: {
        id: 'ab6sof83s1o2',
        title: 'test project 3',
        completed: false,
        completedDeleted: false
      }
    },
    uiState: {
      selectedSection: {
        type: sectionTypes.CONTEXT,
        id: 'a21sogy3s0oq'
      }
    },
    order: {
      project: ['ab6sof83s1o1', 'ab6sof83s1o2'],
      context: ['a21sogy3s0oq', 'a23sogy3s0oq']
    }
  })

  test('Should generate correct props from state', () => {
    const ownProps = { section: 'a21sogy3s0oq'}
    const props = mapStateToProps(testState, ownProps)

    expect(props).toMatchSnapshot()
  })
})
