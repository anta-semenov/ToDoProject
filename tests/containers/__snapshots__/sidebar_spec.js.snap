// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar container Should generate correct props from state 1`] = `
Object {
  "authStatus": "AUTH_NONE",
  "dataStatus": "DATA_NONE",
  "groups": Immutable.List [
    Immutable.Map {
      "type": "BASIC",
      "items": Immutable.List [
        Immutable.Map {
          "type": "inbox",
          "title": "Inbox",
          "active": false,
          "count": 1,
        },
        Immutable.Map {
          "type": "today",
          "title": "Today",
          "active": false,
          "count": 2,
        },
        Immutable.Map {
          "type": "next",
          "title": "Next",
          "active": false,
        },
        Immutable.Map {
          "type": "someday",
          "title": "Someday",
          "active": false,
        },
        Immutable.Map {
          "type": "completed",
          "title": "Completed",
          "active": false,
        },
      ],
    },
    Immutable.Map {
      "type": "CONTEXTS",
      "title": "Contexts",
      "addNewTitle": "Add context",
      "items": Immutable.List [
        Immutable.Map {
          "id": "a21sogy3s0oq",
          "type": "CONTEXT",
          "title": "test context 1",
          "active": true,
          "editing": false,
          "count": 0,
        },
        Immutable.Map {
          "id": "a23sogy3s0oq",
          "type": "CONTEXT",
          "title": "test context 2",
          "active": false,
          "editing": false,
          "count": 1,
        },
      ],
    },
    Immutable.Map {
      "type": "PROJECTS",
      "title": "Projects",
      "addNewTitle": "Add project",
      "items": Immutable.List [
        Immutable.Map {
          "id": "ab6sof83s1o1",
          "type": "PROJECT",
          "title": "test project 2",
          "active": false,
          "editing": false,
        },
        Immutable.Map {
          "id": "ab6sof83s1o2",
          "type": "PROJECT",
          "title": "test project 3",
          "active": false,
          "editing": false,
        },
      ],
    },
  ],
}
`;
